README
IMAP Utils
==========

Collection of small script useful to manage IMAP account.

Homepage
========

<http://cihar.com/software/imap-utils/>

License
=======

GNU GPL version 2.

Configuration
=============

All utils share ~/.imap-utils configuration file. IMAP connection is
configuration section:

[IMAP]
host = hostname ; Host where to connect
login = user ; User name
password = password ; Password
ssl = 1 ; Whether to use SSL
simulate = 1 ; Do not perform actions, only simulate
verbose = 1 ; Print what is going on


imap-purge
==========

Purges old emails from defined folders. Configuration sections:

[purge]
folder = <read>,<unread> ; Purges mails in folder

[purge-recursive]
folder = <read>,<unread> ; Purges mails in folder and sub folders

Read mails are purged after <read> days, unread after <unread> days.


imap-folders
============

Creates folders on IMAP. Configuration sections:

[create]
folder = 1
anotherfolder = 1


imap-stats
==========

Lists number of unread/unseen mails on IMAP.


store-message
=============

Stores message from stdin to folder defined by first argument.


process-messages
================

Reads all messages from folder (first argument) and passes them to
external program (second argument).


Bug reporting
=============

Please report found bugs to <http://bugs.cihar.com/>.


Version control
===============

Developers might be interested in snapshot or version control system
(Subversion is being used).

Anonymous Subversion is available using <svn://svn.cihar.com/imap-utils>,
there is also web browser for it - <http://viewsvn.cihar.com/imap-utils>.

# vim: et ts=4 sw=4 sts=4 tw=72 spell spelllang=en_us


ChangeLog
IMAP Utils
==========

A set of simple utilities for managing IMAP email messages.


News
----

0.6
=== (2015-??-??)

0.5
=== (2015-10-01)

* Changed license to GPL3+.
* Various coding style fixes.

0.4
=== (2008-09-15)

* Improve parsing of folder listing.
* Do not fail on non listable folders.

0.3
=== (2007-08-01)

* Several new scripts for reading and storing messages were added.

0.2
=== (2007-05-24)

* Configuranle listing of folders.
* Support for decoding IMAP names.

0.1
=== (2007-05-15)

* Initial release.


IMAPUtils.py
# -*- coding: UTF-8 -*-
# vim: expandtab sw=4 ts=4 sts=4:
'''
Generic helper class for IMAP-utils.
'''
__author__ = 'Michal Čihař'
__email__ = 'michal@cihar.com'
__license__ = '''
Copyright (c) 2003 - 2007 Michal Čihař

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License version 2 as published by
the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
'''
import sys
import imaplib
import time
import string
import ConfigParser
import os

class IMAPUtils:
    '''
    Wrapper class for IMAP utilities. It takes care about configuration
    and connecting to IMAP.
    '''
    def __init__(self):
        '''
        Reads configuration and prepares private variables defining
        simulation or verbose mode.
        '''
        path = os.path.expanduser('~/.imap-utils')
        self._config = ConfigParser.ConfigParser()
        # We want case sensitive names
        self._config.optionxform = str
        self._config.read(path)
        self._imap = None
        try:
            self._simulate = self._config.getboolean('IMAP', 'simulate')
        except ConfigParser.NoOptionError:
            self._simulate = False
        try:
            self._verbose = self._config.getboolean('IMAP', 'verbose')
        except ConfigParser.NoOptionError:
            self._verbose = False

    def login(self):
        '''
        Logins to IMAP server according to configuration.
        '''
        # Read connection settings
        host = self._config.get('IMAP', 'host')
        login = self._config.get('IMAP', 'login')
        password = self._config.get('IMAP', 'password')
        try:
            ssl = self._config.getboolean('IMAP', 'ssl')
        except ConfigParser.NoOptionError:
            ssl = False

        # Connect to server
        if ssl:
            self._imap = imaplib.IMAP4_SSL(host)
        else:
            self._imap = imaplib.IMAP4(host)

        # Login
        res = self._imap.login(login, password)
        if res[0] != 'OK':
            sys.stderr.write("login: %s\n" % str(res))
            sys.exit(1)

    def get_ints(self, section, name):
        '''
        Reads two ints from configuration.
        '''
        raw = self._config.get(section, name)
        if raw.find(',') == -1:
            return (int(raw), int(raw))
        raw_read, raw_unread = raw.split(',')
        return (int(raw_read), int(raw_unread))

    def get_timestamps(self, section, name):
        '''
        Reads two timestamp deltas from configuration.
        '''
        read, unread = self.get_ints(section, name)
        return (
                time.localtime(time.time() - (read * 86400)) ,
                time.localtime(time.time() - (unread * 86400))
                )



IMAPUtils/Codec.py
# -*- coding: UTF-8 -*-
# vim: expandtab sw=4 ts=4 sts=4:
'''
IMAP modified UTF-7 codec

See https://bugs.python.org/issue5305
'''
__author__ = 'Michal Čihař'
__email__ = 'michal@cihar.com'
__license__ = '''
Copyright (c) 2003 - 2014 Michal Čihař

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import codecs


def modified_base64(s):
    s_utf7 = s.encode('utf-7')
    return s_utf7[1:-1].replace('/', ',')


def modified_unbase64(s):
    s_utf7 = '+' + s.replace(',', '/') + '-'
    return s_utf7.decode('utf-7')


def encoder(s):
    r = []
    _in = []
    for c in s:
        if ord(c) in (range(0x20, 0x26) + range(0x27, 0x7f)):
            if _in:
                r.extend(['&', modified_base64(''.join(_in)), '-'])
                del _in[:]
            r.append(str(c))
        elif c == '&':
            if _in:
                r.extend(['&', modified_base64(''.join(_in)), '-'])
                del _in[:]
            r.append('&-')
        else:
            _in.append(c)
    if _in:
        r.extend(['&', modified_base64(''.join(_in)), '-'])
    return (''.join(r), len(s))


def decoder(s):
    r = []
    decode = []
    for c in s:
        if c == '&' and not decode:
            decode.append('&')
        elif c == '-' and decode:
            if len(decode) == 1:
                r.append('&')
            else:
                r.append(modified_unbase64(''.join(decode[1:])))
            decode = []
        elif decode:
            decode.append(c)
        else:
            r.append(c)
    if decode:
        r.append(modified_unbase64(''.join(decode[1:])))
    return (u''.join(r), len(s))


class StreamReader(codecs.StreamReader):
    def decode(self, s, errors='strict'):
        return decoder(s)


class StreamWriter(codecs.StreamWriter):
    def decode(self, s, errors='strict'):
        return encoder(s)


def imap4_utf_7(name):
    if name == 'imap4-utf-7':
        return (encoder, decoder, StreamReader, StreamWriter)


IMAPUtils/__init__.py
# -*- coding: UTF-8 -*-
# vim: expandtab sw=4 ts=4 sts=4:
'''
Generic helper class for IMAP-utils.
'''
__author__ = 'Michal Čihař'
__email__ = 'michal@cihar.com'
__license__ = '''
Copyright (c) 2003 - 2014 Michal Čihař

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

__version__ = '0.6'

import sys
import imaplib
import time
import ConfigParser
import os
import codecs
from IMAPUtils.Codec import imap4_utf_7

codecs.register(imap4_utf_7)


class IMAPUtil(object):
    '''
    Wrapper class for IMAP utilities. It takes care about configuration
    and connecting to IMAP.
    '''
    def __init__(self):
        '''
        Reads configuration and prepares private variables defining
        simulation or verbose mode.
        '''
        path = os.path.expanduser('~/.imap-utils')
        self._config = ConfigParser.ConfigParser()
        # We want case sensitive names
        self._config.optionxform = str
        self._config.read(path)
        self._imap = None
        try:
            self._simulate = self._config.getboolean('IMAP', 'simulate')
        except ConfigParser.NoOptionError:
            self._simulate = False
        try:
            self._verbose = self._config.getboolean('IMAP', 'verbose')
        except ConfigParser.NoOptionError:
            self._verbose = False

    def login(self):
        '''
        Logins to IMAP server according to configuration.
        '''
        # Read connection settings
        host = self._config.get('IMAP', 'host')
        login = self._config.get('IMAP', 'login')
        password = self._config.get('IMAP', 'password')
        try:
            ssl = self._config.getboolean('IMAP', 'ssl')
        except ConfigParser.NoOptionError:
            ssl = False

        # Connect to server
        if ssl:
            self._imap = imaplib.IMAP4_SSL(host)
        else:
            self._imap = imaplib.IMAP4(host)

        # Login
        res = self._imap.login(login, password)
        if res[0] != 'OK':
            sys.stderr.write("login: %s\n" % str(res))
            sys.exit(1)

    def get_ints(self, section, name):
        '''
        Reads two ints from configuration.
        '''
        raw = self._config.get(section, name)
        if raw.find(',') == -1:
            return (int(raw), int(raw))
        raw_read, raw_unread = raw.split(',')
        return (int(raw_read), int(raw_unread))

    def get_timestamps(self, section, name):
        '''
        Reads two timestamp deltas from configuration.
        '''
        read, unread = self.get_ints(section, name)
        return (
            time.localtime(time.time() - (read * 86400)),
            time.localtime(time.time() - (unread * 86400))
        )
